### PR TITLE
bpo-38863: Improve is_cgi() in http.server

### PR DESCRIPTION
**Pull request by kkangshawn**
is_cgi() function of http.server library does not currently handle a
cgi script if one of the cgi_directories is located at the
sub-directory of given path. Since is_cgi() in CGIHTTPRequestHandler
class separates given path into (dir, rest) based on the first seen
'/', multi-level directories like /sub/dir/cgi-bin/hello.py is divided
into head=/sub, rest=dir/cgi-bin/hello.py then check whether '/sub'
exists in cgi_directories = [..., '/sub/dir/cgi-bin'].
This patch makes the is_cgi() keep expanding dir part to the next '/'
then checking if that expanded path exists in the cgi_directories.

Signed-off-by: Siwon Kang <kkangshawn@gmail.com>

https://bugs.python.org/issue38863

Automerge-Triggered-By: @asvetlov

### CODE REVIEW

**Review by corona10 — COMMENTED** [review]
I now understood what you want to say.
Can you please write the unit test for this?
You can easily provide the test case within inherit `CGIHTTPRequestHandler`

**Review by corona10 — CHANGES_REQUESTED** [review]
Please add a news for changing.
You can follow the below guide.
https://devguide.python.org/committing/#what-s-new-and-news-entries

Thank you for the contribution

File: Lib/http/server.py
Comment (corona10) [inline Lib/http/server.py:1017]: The unit test should be done with appended or overridden `self.cgi_directories`

Comment (kkangshawn) [reply Lib/http/server.py:1017]: Thank you so much for your kind explanation. I will upload my unit test code right away with the news and get back to you

**Comment by corona10** [comment]
@asvetlov  @vadmium,

Dear core developers, 

@kkangshawn is a first-time contributor to this project.
Please guide for his first PR :)

Thanks for your understanding.

File: Misc/NEWS.d/next/Library/2019-11-21-16-30-00.bpo-38863.RkdTjf.rst
Comment (corona10) [inline Misc/NEWS.d/next/Library/2019-11-21-16-30-00.bpo-38863.RkdTjf.rst]:
````
:mod:`http.server`
````

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2019-11-21-16-30-00.bpo-38863.RkdTjf.rst
Comment (kkangshawn) [reply Misc/NEWS.d/next/Library/2019-11-21-16-30-00.bpo-38863.RkdTjf.rst]: I appreciate for your valuable advice. I will get back with the change soon.

**Review by corona10 — APPROVED** [review]
I am okay with the current change, but the final decision will be decided by core developers.
In case, you should discuss the change will be applied or not on https://bugs.python.org/issue38863

I add the @ethanfurman as the cgi expert.

**Comment by kkangshawn** [comment]
Thank you @corona10 for everything you have helped me on my first contribution to the python project! I expect to see you next :)

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks for your contribution.
Since the PR is an enhancement it should be applied to the next Python only.
We have the policy to backport bug fixes only.

**Comment by kkangshawn** [comment]
I get it. Thank you @asvetlov for the review!

**Comment by pablogsal** [comment]
This pull request introduced a reference leak in `test_httpservers`. lease, check out https://bugs.python.org/issue38962 or the refleak buildbots for more info.

Reverting commit 91daa9d7224626dad4bb820924c01b3438ca6e3f fixes the problem.

**Pull request merged**
